Restructuring API documentation (#160)
Altinn/altinn-studio-docs: content/teknologi/altinnstudio/altinn-api/app-api/instances/_index.md

---
title: Instances
description: The App API to work with instances of an App
toc: true
tags: [api]
weight: 50
---

## Overview

> An [instance](../../models/instance) works as a form of envelope or folder where data can be collected and exchanged between the user and owner of the application. The instance document is a way for Altinn and external parties to track the state of one specific data exchange. How long an instance can live and how many interactions there can be between the application owner and user will vary from one App to another. Advanced Apps will have their own documentation.

Reviewer: ```suggestion
An [instance](../../models/instance) can be regarded as an envelope or folder where data is collected and exchanged between the application user and owner. The instance document is a way for Altinn and external parties to track the state of one specific data exchange. How long an instance can live and how many interactions there can be between the application owner and user will vary from one App to another. Advanced Apps will have their own documentation.
```

> An [instance](../../models/instance) works as a form of envelope or folder where data can be collected and exchanged between the user and owner of the application. The instance document is a way for Altinn and external parties to track the state of one specific data exchange. How long an instance can live and how many interactions there can be between the application owner and user will vary from one App to another. Advanced Apps will have their own documentation.

Reviewer: ```suggestion
An [instance](../../models/instance) works as a form of envelope or folder where data can be collected and exchanged between the user and owner of the application. The instance document is a way for Altinn and external parties to track the state of one specific data exchange. How long an instance lives and how many interactions there are
 between the application owner and user will vary from one App to another. Advanced Apps will have their own documentation.
```

**basePath**
```http
{org}/{appname}/instances
```

## Get instance

Endpoint for downloading the instance metadata document for a specific instance. The App API does not have an endpoint for listing instances. The Storage API in the Platform has a query endpoint that can be used for this purpose.

> Use this endpoint for instances that are active, and the owner party id and instance guid for the instance is known. External systems that work with an instance over multiple sessions might want to keep these values (or full URL) stored on their end to limit the need to query this information multiple times.

Reviewer: ```suggestion
Use this endpoint for instances that are active, and the owner party id and instance guid for the instance are known. External systems that work with an instance over multiple sessions might want to keep these values (or full URL) stored on their end to limit the need to query this information multiple times.
```

```http
GET {basePath}/{instanceOwnerPartyId}/{instanceGuid}
```

## Create instance

> Altinn assigns an unique identifier to all users that wishes to report data. We call this id *instanceOwner.partyId*. 

Reviewer: ```suggestion
Altinn assigns a unique identifier to all users that wishes to report data. We call this id *instanceOwner.partyId*. 
```

Going of the sound of the word rather than the spelling.

If you do not know this, you should provide the official identity number, e.g national identification number for persons or organisation number for organisations,
and in some case user name. This should be provided as part of the payload to the creation request.

> Altinn will look up this identifier and replace it with the instanceOwner.PartyId. The official identity number will be stored in the instance metadata.

Reviewer: ```suggestion
Altinn will look up this identifier and replace it with the instanceOwner.partyId. The official identity number will be stored in the instance metadata.
```

Data elements can be provided as part of the creation request, but can also be uploaded at a later time.

> The client specify the instance owner and can set a number of the metadata fields of the instance by attaching the following form:

Reviewer: ```suggestion
The client specifies the instance owner and can set a number of the metadata fields of the instance by attaching the following form:
```

> The client specify the instance owner and can set a number of the metadata fields of the instance by attaching the following form:

Reviewer: ```suggestion
The client specify the instance owner and may set a number of the metadata fields of the instance by attaching the following form:
```

```json
{
    "appId" : "org/app",
    "instanceOwner": {
        "personNumber": "12247918309",
        "organisationNumber": null
    },
    "dueBefore": "2019-06-01T12:00:00Z",
    "visibleAfter": "2019-05-20T00:00:00Z"
}
```

Notice that all dates must be expressed in **Utc (Zulu)** time zone and represented according to ISO 8601!

> Data elements (files) can be attached to the initial request as a *multipart/form-data* or as *attachments*.

Reviewer: multipart/form-data is referred to as multipart form-data futher down.  Can we use the same expression both places, not sure which one is more correct.

The name of the parts must correspond to element types defined in the application metadata. 

```http
POST {basePath}
```

![Flow chart for instantiation](instantiate-for-an-instance-owner.png "Instantiate for instance owner")

A multipart formdata should contain the instance json document and the data element files of the instance.
The first part should be *instance* which contains the json template to create an instance from.

The subsequent parts must have a name that correspond to the element types defined in application metadata.
They may have a filename. Hence the *model1* and *certificate* names correspond to data types defined in the application metadata.

> If more data elements are needed they must be defined in the application metadata.

Reviewer: ```suggestion
If additional data elements are required, they must be defined in the application metadata.
```

```http {linenos=false,hl_lines=[5,10,15]}
Content-Type: multipart/form-data; boundary="abcdefg"

--abcdefg
Content-Type: application/json; charset=utf-8
Content-Disposition: form-data; name="instance"
{ ... }

--abcdefg
Content-Type: application/xml
Content-Disposition: form-data; name="model1"
<xml> ... </xml>

--abcdefg
Content-Type: application/pdf
Content-Disposition: form-data; name="certificate"; filename=certificate.pdf
%PDF-1.4
%Óëéá
1 0 obj
...

--abcdefg--
```

This call will return the instance metadata document that was created. 

## Complete instance

> Endpoint used by Application Owner to mark an instance as completed. Technically this only means that the instance is no longer needed by the Application Owner. Altinn can delete the instance and associated data should the Application user decide to do so.

Reviewer: ```suggestion
Endpoint used by application owner to mark an instance as completed. Technically this only means that the instance is no longer needed by the application owner. Altinn can delete the instance and associated data should the Application user decide to do so.
```

Not capitalized elsewhere on this page

Reviewer: Seems to be a mix and match.. As long as we are consistent whichever is fine by me.

> Endpoint used by Application Owner to mark an instance as completed. Technically this only means that the instance is no longer needed by the Application Owner. Altinn can delete the instance and associated data should the Application user decide to do so.

Reviewer: ```suggestion
Endpoint used by Application Owner to mark an instance as completed. Technically this only means that the instance is no longer needed by the Application Owner. Altinn can delete the instance and associated data should the instance owner decide to do so.
```
How about referring to the instanceOwner here as a way to link it back to the url ? might be confusing as someone elese may represent the  instance owner?

> Endpoint used by Application Owner to mark an instance as completed. Technically this only means that the instance is no longer needed by the Application Owner. Altinn can delete the instance and associated data should the Application user decide to do so.

Reviewer: Could we link the reasoning behind deletion closer to the complete action? This way it would make more sense that this information is included here.

E.g. When both application owner and instance owner have indicated that they no longer need the instance, through completion or deletion, Altinn will delete the instance and associated data.

```http
POST {basePath}/{instanceOwnerPartyId}/{instanceGuid}/complete
```

The request does not use the request body for anything even though it is a POST request.

## Update sub status

The instance [sub status](https://altinn.github.io/docs/altinn-studio/app-creation/instance/#substatus) is used to give an end user further details about the state of their instance. Currently, only application owner is allowed to update substatus for an instance. Include the new substatus in the body of the requests as a json.

```http
PUT {basePath}/{instanceOwnerPartyId}/{instanceGuid}/substatus
```

**Request body**
```json
{
    "label": "some.label",
    "description": "Beskrivelse i klarteskst"
}
```
The values can be referencing text resource names from the language files or be regular text.

## Delete instance

> Endpoint for marking an instance as deleted. This can be used by both Application Owner and User to delete an instance. The endpoint has an optional parameter called **hard** that can be used to indicate the type of delete that is wanted. Hard delete means that the instance will be unrecoverable and deleted completely by Altinn within a few days. Setting the parameter to false or simply omitting it, will move the instance to a recycle bin. 

Reviewer: Deleted within a few days is a truth with modifications I guess.. Any way to communicate that there are additional factors that affect the actual deletion time?

Reviewer: Or are we assuming that only drafts are deleted using this endpoint? Befause in that case, the statement would be accurate.

```http
DELETE {basePath}/{instanceOwnerPartyId}/{instanceGuid}?hard=true
```

The endpoint does not use the request body for anything. 

There are no Application API endpoint for recovering instances in the recycle bin.

> ## Instance read status

Reviewer: How about making Update read status the section header and combining the two paragraphs? 
I would also consider grouping this section with the sub-status.
